fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java

```diff
@@ -0,0 +1,384 @@
+    public static final ImmutableList<String> WAREHOUSE_PROC_NODE_TITLE_NAMES = new ImmutableList.Builder<String>()
+            .add("Warehouse")
+            .add("State")
+            .add("Size")
+            .add("min_cluster")
+            .add("max_cluster")
+            .add("cluster_count")
+            .add("total_pending")
+            .add("total_running")
+            .build();
+
+    private void readLock() {
+        this.rwLock.readLock().lock();
+    }
+    private void readUnlock() {
+        this.rwLock.readLock().unlock();
+    }
+
+    private void writeLock() {
+        this.rwLock.writeLock().lock();
+    }
+
+    private void writeUnLock() {
+        this.rwLock.writeLock().unlock();
+    }
+
+    public WarehouseManager() {
+    }
+
+    public Warehouse getWarehouse(String warehouseName) {
+        return fullNameToWh.get(warehouseName);
+    }
+
+    public boolean warehouseExists(String warehouseName) {
+        readLock();
+        try {
+            return fullNameToWh.containsKey(warehouseName);
+        } finally {
+            readUnlock();
+        }
+    }
+
+    // these apis need lock protection
+    public void createWarehouse(CreateWarehouseStmt stmt) throws DdlException {
+        createWarehouse(stmt.getFullWhName(), stmt.getProperties());
+    }
+
+    public void createWarehouse(String whName, Map<String, String> properties) {
+        readLock();
+        try {
+            Preconditions.checkState(!fullNameToWh.containsKey(whName), "Warehouse '%s' already exists", whName);
+        } finally {
+            readUnlock();
+        }
+
+        writeLock();
+        try {
+            Preconditions.checkState(!fullNameToWh.containsKey(whName), "Warehouse '%s' already exists", whName);
+            long id = GlobalStateMgr.getCurrentState().getNextId();
+            Warehouse wh = new Warehouse(id, whName, properties);
+            fullNameToWh.put(wh.getFullName(), wh);
+            idToWh.put(wh.getId(), wh);
+            wh.setExist(true);
+            GlobalStateMgr.getCurrentState().getEditLog().logCreateWh(wh);
+            LOG.info("createWarehouse whName = " + whName + ", id = " + id);
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void replayCreateWarehouse(Warehouse warehouse) {
+        String whName = warehouse.getFullName();
+        readLock();
+        try {
+            Preconditions.checkState(!fullNameToWh.containsKey(whName), "Warehouse '%s' already exists", whName);
+        } finally {
+            readUnlock();
+        }
+
+        writeLock();
+        try {
+            fullNameToWh.put(whName, warehouse);
+            idToWh.put(warehouse.getId(), warehouse);
+            warehouse.setExist(true);
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void suspendWarehouse(SuspendWarehouseStmt stmt) throws DdlException {
+        String whName = stmt.getFullWhName();
```

Comment:
I think we should put this entire function in write lock and we should log in write lock or else there will be some consistency problem which may cause the follower crash when replaying log.

Reply:
resumeWarehouse also has this problem